core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala

```diff
@@ -283,6 +286,48 @@ class ProduceRequestTest extends BaseRequestTest {
     assertEquals(-1, partitionProduceResponse1.logAppendTimeMs)
   }
 
+  /**
+   * See `ApiKeys.PRODUCE_OLDEST_VERSION` for the details of why we need special handling for produce request v0-v2 (inclusive).
+   */
+  @Test
+  def testProduceRequestV0ToV2IsRejectedByBroker(): Unit = {
```

Comment:
> We would have to remove testProduceRequestV0ToV2IsRejectedByBroker since there would be no way to produce with v0-v2. We would probably have to create a ducktape test to replace that. I am not sure how easy or hard that would be (ducktape tests typically use clients instead of testing specific requests).

We removed a bunch of old version of other requests. If we just follow their approach by removing the old version from the protocol definition, there is no need to test the server rejection logic just for produce request, right?

Reply:
This one is different because we have the overrides - it's very difficult to ensure we don't regress as the code changes without a test. For the other cases, the logic is very simple.

Reply:
It seems the primary purpose of retaining v0-v2 is to facilitate easier test writing. Perhaps we could create a `LegacyProduceRequest` json that mirrors the v0-v2 format exclusively for testing purposes?

Reply:
That won't work, the separate file would not have be able to use the same api key id.

Reply:
> the separate file would not have be able to use the same api key id.

Pardon me, what is the issue of defining `"apiKey": 0` in other file (test scope) ? I create `kafka/clients/src/test/resources/common/message/LegacyProduceRequest.json` according to v0-v2, and it works well.

Reply:
I hadn't considered doing this in a different scope. I can try that and see if it works.

Reply:
I realized that there was a simple way - described it in the main PR.